modules/gateways/veritrans.php

```diff
@@ -349,21 +349,54 @@ function veritrans_link($params)
         $params['credit_card']['save_card'] = true;
     }
 
-    // Get snap token
     try {
-        // debug
-        // echo "<pre>";
-        // var_dump($params);
-        // echo "</pre>";
-        // exit();
-        $snap_transaction = Veritrans_Snap::createTransaction($params);
-        $snapToken = $snap_transaction->token;
-        $redirect_url = $snap_transaction->redirect_url;
-        // error_log(" ############# TOKEN ::: ".$snapToken);
+        $whmcsApiCommand = 'GetInvoice';
+        $whmcsApiPostParams = array(
+            'invoiceid' => $invoiceId
+        );
+        $invoice = localAPI($whmcsApiCommand, $whmcsApiPostParams);
+        $existingInvoiceNotes = $invoice->notes;
     } catch (Exception $e) {
-        // error_log('Caught exception: ',  $e->getMessage(), "\n");
-        if (preg_match('/utilized/',$e->getMessage()) ){
-            return "<h4> Awaiting Your Payment </h4>";
+        $existingInvoiceNotes = "";
+    }
+
+    $midtransNotesSeparator = "#PayUrl: ";
+    // Check if Snap URL/token found in invoices notes, then use the existing Snap URL/token
+    if (strpos($existingInvoiceNotes, $midtransNotesSeparator)) {
+        $redirect_url = substr($existingInvoiceNotes, strpos($existingInvoiceNotes, 'midtrans.com'));
+        $extractToken = explode('/', $existingInvoiceNotes);
+        $snapToken = array_pop($extractToken);
+    } else {
+        // Get snap token
+        try {
+            // debug
+            // echo "<pre>";
+            // var_dump($params);
+            // echo "</pre>";
+            // exit();
+            $snap_transaction = Veritrans_Snap::createTransaction($params);
+            $snapToken = $snap_transaction->token;
+            $redirect_url = $snap_transaction->redirect_url;
+            // error_log(" ############# TOKEN ::: ".$snapToken);
+
+            // save snap redirect url to invoice notes
+            try {
+                $whmcsApiCommand = 'UpdateInvoice';
+                $whmcsApiPostParams = array(
+                    'invoiceid' => $invoiceId,
+                    'notes' => $existingInvoiceNotes . $midtransNotesSeparator . $redirect_url
```

Comment:
After some consideration, better modify the UpdateInvoice value to also have end separator, like this:
```
'notes' => $existingInvoiceNotes . $midtransNotesSeparatorBegin . $redirect_url . $midtransNotesSeparatorEnd
```
with
```
    $midtransNotesSeparatorBegin = "#PayUrl: ";
    $midtransNotesSeparatorEnd = " PayUrl#";
```